feat: `DC` dialect operators

luisacicolini opened the pull request.

tobiasgrosser committed:
feat: add DC dialect

alexkeizer committed:
DC: semantics for dc.merge

Co-authored-by: Yann Herklotz <yann@yannherklotz.com>

alexkeizer committed:
rename `dropFirst` to `tail`, to be consistent with stream terminology

alexkeizer committed:
define semantics for a fully determinate, alternating merge

alexkeizer committed:
module docstring

alexkeizer committed:
cleanup, and add a comment about deterministic vs determinate

alexkeizer committed:
reorganize file into sections

alexkeizer committed:
rename `Brook` to `DC.Stream`, and factor the basic definitions out to separate file

alexkeizer committed:
WIP: weak bisimulation for sparse streams

Commit:
expanded to more types - tbc

Commit:
added some more ops

Commit:
bisim streams

Commit:
bisim streams

ymherklotz committed:
Update lake dependencies for 4.8.0

ymherklotz committed:
Add some definitions to WeakBisim

ymherklotz committed:
Add determinate theorem

ymherklotz committed:
Prove determinate theoremm

luisacicolini committed:
potential progress in sight

luisacicolini committed:
pairing

luisacicolini committed:
example attempt to fixed but failed

luisacicolini committed:
init dc and separation from handshake

luisacicolini committed:
merged bisigm branch
Co-authored-by: Yann Herklotz <yann@yannherklotz.com>

luisacicolini committed:
merge main

luisacicolini committed:
repaired conflict

luisacicolini committed:
started join

luisacicolini committed:
removed old DC

luisacicolini committed:
fixed new DC semantics

luisacicolini committed:
fixed handshake

luisacicolini committed:
almost done fixing bisim

luisacicolini committed:
bisim won't win

luisacicolini committed:
bisim updated with stream taking arg

luisacicolini committed:
removed useless import

luisacicolini committed:
tried to fix example and failed miserably

luisacicolini committed:
merge main

luisacicolini committed:
merged

luisacicolini committed:
fixing example

luisacicolini committed:
result of pairing, cc @yann

luisacicolini committed:
join handshake op

luisacicolini committed:
mux

luisacicolini committed:
fixed join and mux

luisacicolini committed:
sink, source, sync

ymherklotz committed:
Add the definition of the Token

luisacicolini committed:
merge main

luisacicolini committed:
dc stream ops

luisacicolini committed:
dialect structure

luisacicolini committed:
merge main

luisacicolini committed:
denotation

luisacicolini committed:
ready for pr

luisacicolini committed:
removed unnecessary edit to forlean

luisacicolini committed:
removed MLIR2DC which is temp

luisacicolini committed:
typo

luisacicolini committed:
removed lakefile

luisacicolini committed:
removed MLIR2DC

ymherklotz commented on SSA/Projects/CIRCT/DC/DC.lean:
```diff
@@ -0,0 +1,184 @@
+
+-- weirdly enough the docs says that c must be a Value type (not token),
+-- but I don't see how this would be useful? I left stream bool now, but I don't like it that much
+-- emotionally not a valuestream but can do for now
```
Hmm the docs say that `c` is a ValueStream (DC.value) right?  Which is what we want for the branch, so I think that the implementation matches the documentation/specification in this case, which is good.  It can't be a TokenStream because otherwise you wouldn't have a condition to branch on.  So I think this implementation is exactly the one specified by the specification.

ymherklotz commented on SSA/Projects/CIRCT/DC/DC.lean:
```diff
@@ -0,0 +1,184 @@
+    | none, none => (none, (x.tail, y.tail))
+
+-- same issue as branch for the condition stream
+def select (x y : TokenStream) (c : Stream Bool): TokenStream :=
```
I would just change `Stream Bool` to `ValueStream Bool`, but otherwise my comment from above works here too, and I think this matches the documentation actually.

ymherklotz commented on SSA/Projects/CIRCT/DC/DC.lean:
```diff
@@ -0,0 +1,184 @@
+def sink (x : TokenStream) : TokenStream :=
+  Stream.corec (β := TokenStream) x fun x => (none, x.tail)
+
+def source (a : Unit) : TokenStream :=
```
I don't think this needs an input `a`, it can just generate an infinite stream of `some ()`.

ymherklotz commented on SSA/Projects/CIRCT/DC/DC.lean:
```diff
@@ -0,0 +1,184 @@
+    | none, none, some _ => (none, (x.tail, y.tail, c))
+
+def sink (x : TokenStream) : TokenStream :=
+  Stream.corec (β := TokenStream) x fun x => (none, x.tail)
```
Yep I agree with the implementation of sink.

ymherklotz commented on SSA/Projects/CIRCT/DC/DC.lean:
```diff
@@ -0,0 +1,184 @@
+    | none, some _ => (none, (x.tail, y))
+    | none, none => (none, (x.tail, y.tail))
+
+def merge (x y : TokenStream) : TokenStream :=
```
I think the type should be `TokenStream -> TokenStream -> ValueStream Bool`, where the value stream says true or false depending on if the left or right token was consumed.

ymherklotz commented on SSA/Projects/CIRCT/DC/DC.lean:
```diff
@@ -0,0 +1,184 @@
+    fun x => Id.run <| do
+      (x 0, x 0, x.tail)
+
+def join (x y : TokenStream) : TokenStream  :=
```
For now this looks good, but maybe we need to see if we can implement variadic arguments at some point, but for now we don't have to focus on that yet.

ymherklotz commented on SSA/Projects/CIRCT/DC/DC.lean:
```diff
@@ -0,0 +1,184 @@
+abbrev ValueStream := Stream
+abbrev TokenStream := Stream Unit
+
+def unpack (x : ValueStream α) : ValueStream α × TokenStream :=
```
I think for now I'm happy with these semantics.  We are maybe missing out on a bit of the benefit of DC, because we are not really separating data and control, but I think we could refine it a bit later on.  A solution that I have started to think about is to really have an explicit separation between data and control.  Essentially, you can only have data operations between an `unpack` and a `pack`, and so you could just treat those regions differently.  Maybe we can explore that a bit, but at the same time we can use these semantics and try and prove an initial lowering from Handshake to DC.

luisacicolini replied:
Yes I agree with this separation and I think the semantics we have so far is a good starting point for that, since `TokenStream` is explicitly different than `ValueStream` we can just use the two for different operations (control vs. data)

ymherklotz commented on SSA/Projects/CIRCT/Handshake/Handshake.lean:
```diff
@@ -112,11 +115,71 @@ def altMerge (x y : Stream α) : Stream α :=
+      | _, some _, some false => (y 0, (x, y.tail, c.tail)) -- pop from y
+      | _, _, none => (none, (x, y, c.tail)) -- no pop
+
+-- discards any data: actually this should not return anything
```
Yeah in practice this should not return anything, but to be able to consume a stream you have to produce another stream.  But as long as the stream never blocks and can always consume a stream, then I think it has the behaviour that we are looking for.

ymherklotz commented on lakefile.lean:
I'm not sure why this file is part of this commit, should it be?

luisacicolini committed:
dialect structure more

luisacicolini committed:
tmp

luisacicolini committed:
solved comments

luisacicolini committed:
dc structure

luisacicolini committed:
updated ci because it was driving me crazy

luisacicolini committed:
forgot folder in ci

luisacicolini committed:
sorried unsolved goals in wip handshake example

luisacicolini committed:
fixed sorry

luisacicolini committed:
tentative example (does not work yet)

luisacicolini committed:
new semantics for branch, i think this is more accurate wrt docs

luisacicolini committed:
a simple example that works

luisacicolini committed:
fixed ci

luisacicolini commented:
Added a small working example. I tried adding a more complex one using `DC.branch`, but somehow it's complaining about having two outputs, so for now we have this. I also changed the definition of `branch` slightly, this one seems to me more compliant with the docs.

luisacicolini committed:
fixed filenames

luisacicolini committed:
new branch example

ymherklotz reviewed (approved):
Looks good to me now, this implements all DC operations with pure streaming semantics and adds the WeakBisim relation to compare streams.

luisacicolini commented:
Happy to merge unless there are other issues CC: @tobiasgrosser @alexkeizer
